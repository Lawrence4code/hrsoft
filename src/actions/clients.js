import uuid from 'uuid';
import database from '../firebase/firebase';
// action generator for adding client
// passed with default value if no value passed for the client
export const addClient = client => ({
  type: 'ADD_CLIENT',
  client
});

export const startAddClient = (clientData = {}) => {
  return dispatch => {
    const {
      title = '',
      firstName = '',
      lastName = '',
      gender = '',
      phone = '',
      email = '',
      address = '',
      status = '',
      field = '',
      note = '',
      lastCommuniation = '',
      memberType = '',
      createdAt = 0
    } = clientData;

    const client = {
      title,
      firstName,
      lastName,
      gender,
      phone,
      email,
      address,
      status,
      field,
      note,
      lastCommuniation,
      memberType,
      createdAt
    };

    return database
      .ref('clients')
      .push(client)
      .then(ref => {
        dispatch(
          addClient({
            id: ref.key,
            ...client
          })
        );
      });
  };
};

// action generating for removing client
export const removeClient = ({ id } = {}) => {
  return {
    type: 'REMOVE_CLIENT',
    id
  };
};

// action generating for modifying the client details
export const editClient = (id, updates) => {
  return {
    type: 'EDIT_CLIENT',
    id,
    updates
  };
};
