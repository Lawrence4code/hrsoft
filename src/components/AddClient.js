import React from 'react';
import { Link } from 'react-router-dom';

const AddClient = () => (
  <div>
    <h2>AddClientPage</h2>
    <Link to="/"> Back to Dashboard </Link>
  </div>
);

export default AddClient;
