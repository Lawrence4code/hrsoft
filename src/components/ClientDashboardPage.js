import React from 'react';
import { Link } from 'react-router-dom';
import ClientList from './ClientList';
import ClientListFilters from './ClientListFilters';
import ClientSummary from './ClientSummary';

const ClientDashboardPage = () => (
  <div>
    <ClientSummary />
    <ClientListFilters />
    <ClientList />
  </div>
);

export default ClientDashboardPage;
