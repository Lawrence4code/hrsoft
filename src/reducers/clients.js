// default values for the cient reducer
const clientReducerDefaultState = [];

const clientsReducer = (state = clientReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_CLIENT':
      return [...state, action.client];
    case 'EDIT_CLIENT':
      return state.map(client => {
        if (client.id === action.id) {
          return {
            ...client,
            ...action.updates
          };
        } else {
          return client;
        }
      });
    case 'REMOVE_CLIENT':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};

export default clientsReducer;
