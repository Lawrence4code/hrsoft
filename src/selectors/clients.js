// get visible clients
import moment from 'moment';

const getVisibleClients = (clients, { name, sortBy, startDate, endDate }) => {
  // console.log(clients);
  // console.log(startDate, endDate);
  return clients
    .filter(client => {
      const createdAtMoment = moment(client.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;

      const firstNameMatch = client.firstName
        .toLowerCase()
        .includes(name.toLowerCase());
      const lastNameMatch = client.lastName
        .toLowerCase()
        .includes(name.toLowerCase());

      return (
        startDateMatch && endDateMatch && (firstNameMatch || lastNameMatch)
      );
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'alphabetically') {
        console.log('alpha sort');
        return -1;
      }
    });
};

export default getVisibleClients;
