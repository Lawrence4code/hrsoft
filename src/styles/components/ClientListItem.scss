.clientListItems {
  & div {
    color: #636363;
    font-size: 14px;

    border-radius: 3px;
  }

  & div:nth-child(odd) {
    background: #ececec;
  }
  & div:nth-child(even) {
    background: #ffffff;
  }

  &__message {
    margin-top: 1rem;
    text-align: center;
    color: #7a7d82;
  }
}

.clientListItem {
  display: grid;
  grid-template-columns: 1.55fr 1.85fr 1.25fr 80px 1fr 100px 100px 100px 100px;
  text-align: center;
  align-content: center;
  margin: 0 auto;
  width: 90%;
  padding: 0.5rem 0 0.5rem 0;

  & > * {
    align-self: center;
  }

  &__name {
    text-transform: capitalize;
    text-decoration: none;
    color: #4c4c4c;
    font-weight: bold;
    align-content: center;
    display: grid;
    grid-template-columns: 1fr 5px;
  }

  &__email {
    text-transform: lowercase;
  }

  &__memberType {
    text-transform: capitalize;
  }

  &__profession {
    text-transform: capitalize;
  }

  &__actions {
    display: flex;
    justify-content: center;
    align-items: center;
  }
}

.binLogo {
  width: 1rem;
  cursor: pointer;
  margin: 0 8px;
}

.editLogo {
  width: 1.3rem;
  cursor: pointer;
  background-color: #8596fc;
  border-radius: 50%;
  padding: 3px;
}

.readMore {
  width: 1.2rem;
  cursor: pointer;
  padding: 3px;
}
