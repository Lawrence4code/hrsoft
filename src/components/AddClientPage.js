import React from 'react';
import ClientForm from './ClientForm';
import { connect } from 'react-redux';
import { startAddClient } from '../actions/clients';

export class AddClientPage extends React.Component {
  onSubmit = client => {
    this.props.startAddClient(client);
    this.props.history.push('/');
  };

  render() {
    return (
      <div className="addClientPage">
        <div className="addClientPage__heading">
          <div className="addClientPage__content">
            <h1> Add New Client </h1>
          </div>
        </div>

        <ClientForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddClient: client => dispatch(startAddClient(client))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddClientPage);
