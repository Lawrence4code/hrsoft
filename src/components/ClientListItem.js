import React from 'react';
import { connect } from 'react-redux';
import { removeClient } from '../actions/clients';
import { Link } from 'react-router-dom';

export const ClientListItem = ({ firstName, lastName, id, dispatch }) => {
  // console.log(props);
  return (
    <div className="clientListItem">
      <Link to={`/edit/${id}`} className="clientListItem__name">
        <p>
          {firstName} {lastName}
        </p>
      </Link>
      <div>
        <a
          className="clientListItem__button clientListItem__button--one"
          onClick={() => {
            dispatch(removeClient({ id }));
          }}
        >
          Remove
        </a>
      </div>

      <Link
        to={`/edit/${id}`}
        className="clientListItem__button clientListItem__button--two"
      >
        Edit
      </Link>
    </div>
  );
};

export default connect()(ClientListItem);
