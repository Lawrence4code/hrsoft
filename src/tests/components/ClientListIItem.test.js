import React from 'react';
import { shallow } from 'enzyme';
import { ClientListItem } from '../../components/ClientListItem';
import clients from '../fixtures/clients';

test('should render ClientListItem', () => {
  const wrapper = shallow(<ClientListItem {...clients[0]} />);
  expect(wrapper).toMatchSnapshot();
});
