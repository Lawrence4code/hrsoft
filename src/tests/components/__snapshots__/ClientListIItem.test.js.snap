// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ClientListItem 1`] = `
<div
  className="clientListItem"
>
  <Link
    className="clientListItem__name"
    replace={false}
    to="/view/1"
  >
    <p>
      John
       
      Doe
    </p>
  </Link>
  <p>
     
    John@doe.com
  </p>
  <p>
     
    987654321
  </p>
  <p>
     
    M
  </p>
  <p>
     
    FakeOne High Street
  </p>
  <p>
     
    Employed
  </p>
  <p>
     
    Designer
  </p>
  <p>
     
    Gold
  </p>
  <p>
     
    N/A
  </p>
  <a
    className="clientListItem__button--one"
    onClick={[Function]}
  >
    Remove
  </a>
  <Link
    className="clientListItem__button clientListItem__button--two"
    replace={false}
    to="/edit/1"
  >
    Edit
  </Link>
</div>
`;
